bin/reflective_loader/src/loader.rs

```diff
@@ -0,0 +1,844 @@
+            panic!("Failed to allocate memory. Last error returned: {}", error_code);
+        }
+    }
+    buffer_handle
+}
+
+// pub unsafe fn GetProcAddress(hmodule: P0, lpprocname: P1) -> FARPROC
+fn get_proc_address(fn_ptr: FnGetProcAddress, err_ptr: FnGetLastError, h_module: isize, lp_proc_name: *const u8) -> isize {
+    let proc_handle = match unsafe { fn_ptr(h_module, lp_proc_name) } {
+        Some(local_proc_handle) => local_proc_handle,
+        None => {
+            let error_code = unsafe { err_ptr() };
+            panic!("Failed to find function {:?} in module. Last error returned: {}", lp_proc_name, error_code);
+        }
+    };
+    proc_handle as isize
+}
+
+
+// pub unsafe fn LoadLibraryA(lplibfilename: P0) ->  Result<HMODULE>
+fn load_library_a(fn_ptr: FnLoadLibraryA, err_ptr: FnGetLastError, lplibfilename: PCSTR) -> HINSTANCE {
+    let library_handle = unsafe { fn_ptr(lplibfilename) };
+    if library_handle == 0 {
+        let error_code = unsafe { err_ptr() };
+        if error_code != 0 {
+            panic!("Failed to load library. Last error returned: {}", error_code);
+        }
+    }
+    library_handle
+}
+
+#[derive(Debug, Copy, Clone)]
+struct BaseRelocationEntry {
+    offset: u16,
+    reloc_type: u16,
+}
+
+impl BaseRelocationEntry {
+    fn new(c_bytes: u16) -> Self {
+        let reloc_type_bit_mask: u16 = 0b1111_0000_0000_0000;
+        let reloc_type = (c_bytes & reloc_type_bit_mask) >> 12;
+        
+        let offset_bit_mask: u16 = 0b0000_1111_1111_1111;
+        let offset = c_bytes & offset_bit_mask;
+        
+        Self {
+            offset,
+            reloc_type,
+        }
+    }
+
+    fn c_size() -> usize {
+        core::mem::size_of::<u16>()
+    }
+}
+
+#[cfg(target_arch = "x86_64")]
+struct PeFileHeaders64 {
+    dos_headers: IMAGE_DOS_HEADER,
+    nt_headers: IMAGE_NT_HEADERS64,
+    section_headers: [IMAGE_SECTION_HEADER; MAX_PE_SECTIONS],
+}
+
+// Pares the PE file from a series of bytes
+#[cfg(target_arch = "x86_64")]
+impl PeFileHeaders64 {
+    fn new(dll_bytes_ptr: *mut c_void) -> Self{
+        // DOS Headers
+        let dos_headers_base_ptr = dll_bytes_ptr as usize;
+        let dos_headers = unsafe { *((dos_headers_base_ptr) as *mut IMAGE_DOS_HEADER) };
+        if dos_headers.e_magic != PE_MAGIC {
+            panic!("PE Magic header mismatch. Expected 0x5A4D == MZ == 21117. File does not appear to be a PE executable.");
+        }
+
+        // NT Headers
+        let nt_headers_base_ptr = dos_headers_base_ptr + dos_headers.e_lfanew as usize;
+        let nt_headers = unsafe { *((nt_headers_base_ptr) as *mut IMAGE_NT_HEADERS64) };
+        if nt_headers.Signature != NT_SIGNATURE {
+            panic!("NT Signature mismatch. Expected 0x4550 == PE == 17744. File does not appear to be a PE executable.");
+        }
+
+        // Section Headers - hopefully there isn't more than MAX_PE_SECTIONS sections.
+        let null_section = IMAGE_SECTION_HEADER{ 
```

Comment:
maybe easier to do `core::mem::zeroed()` instead of defining every field to zero. Unsure if `IMAGE_SECTION_HEADER` has a `Default` impl... that would also work.

Reply:
That's pretty cool.
Only hesitations seems like it adds an unsafe.
Reading the docs seems like there's only undefined behavior when the struct has pointers or refences.
https://doc.rust-lang.org/beta/core/mem/fn.zeroed.html

<img width="984" alt="image" src="https://github.com/KCarretto/realm/assets/7121375/2eb8676a-7822-422a-b89a-b3bf639c808a">

Reply:
Added ✅
The improved readability is nice.

Reply:
Gonna revert for now.
This pulls in a call to `memset` which doesn't exist with `no_std`.
If we can get `compiler-builtins-mem` working i'll revisit.

Reply:
- Blocking on https://github.com/KCarretto/realm/pull/205#discussion_r1243099512